dao/src/main/java/com/epam/esm/dao/impl/CertificateDaoImpl.java

```diff
@@ -1,102 +1,93 @@
+        CriteriaQuery<Certificate> criteriaQuery = queryProvider.buildQuery(searchParameter);
+        return (long) entityManager.createQuery(criteriaQuery).getResultList().size();
     }
 
     @Override
-    public List<Certificate> searchByParameters(String query, List<String> parameters) {
-        String searchQuery = SELECT_BY_SEARCH_PARAMETERS.concat(query);
-        return jdbcTemplate.query(searchQuery, certificateMapper, parameters.toArray());
+    public Long getEntitiesCount() {
+        CriteriaBuilder builder = entityManager.getCriteriaBuilder();
+        CriteriaQuery<Long> query = builder.createQuery(Long.class);
+        query.select(builder.count(query.from(Certificate.class)));
+        return entityManager.createQuery(query).getSingleResult();
     }
 
     @Override
-    public Long create(Certificate entity) {
-        KeyHolder keyHolder = new GeneratedKeyHolder();
-        try {
-            jdbcTemplate.update(connection -> {
-                PreparedStatement ps = connection.prepareStatement(INSERT_CERTIFICATE, new String[]{ColumnName.COLUMN_ID});
-                ps.setString(1, entity.getName());
-                ps.setString(2, entity.getDescription());
-                ps.setBigDecimal(3, entity.getPrice());
-                ps.setInt(4, entity.getDuration());
-                return ps;
-            }, keyHolder);
-        } catch (Exception e) {
-            throw new WrongInsertDataException();
-        }
-        return Objects.requireNonNull(keyHolder.getKey()).longValue();
+    public Optional<Certificate> create(Certificate entity) {
+        entityManager.persist(entity);
+        return Optional.of(entity);
     }
 
     @Override
     public Optional<Certificate> readById(Long id) {
-        return jdbcTemplate.query(SELECT_CERTIFICATE_BY_ID, certificateMapper, id)
-                .stream()
-                .findFirst();
+        return Optional.ofNullable(entityManager.find(Certificate.class, id));
     }
 
     @Override
-    public List<Certificate> readAll() {
-        return jdbcTemplate.query(SELECT_ALL_CERTIFICATES, certificateMapper);
+    public List<Certificate> readAll(int page, int size) {
+        CriteriaQuery<Certificate> query = entityManager.getCriteriaBuilder().createQuery(Certificate.class);
+        Root<Certificate> root = query.from(Certificate.class);
+        query.select(root);
+        return entityManager.createQuery(query)
+                .setFirstResult(getStartPosition(page, size))
+                .setMaxResults(size)
+                .getResultList();
     }
 
     @Override
-    public Integer deleteById(Long id) {
-        return jdbcTemplate.update(DELETE_CERTIFICATE, id);
+    public void deleteById(Long id) {
+        Certificate certificate = entityManager.find(Certificate.class, id);
+        if (entityManager.contains(certificate)) {
+            entityManager.remove(certificate);
+        } else {
+            throw new EntityNotInDBException(id.toString());

```

Comment:
It looks like as a kind of validation and, I believe, service layer is the right place for it. Also is such logic applied to update?

Reply:
right, validation before deleting moved to the service layer